Documentation/auto-dependency-flow/README.md

```diff
@@ -0,0 +1,92 @@
+
+*Full product*: a set of repository build outputs that are coherent. As opposed to "isolated products" which have uncoordinated dependency versions that may conflict.
+
+*Isolated build*: a build that a repo performs independently. The repo chooses its own dependencies, usually pinning to specific versions to isolate itself from breaking changes until they can be taken. This type of build is performed for short-term productivity, and it doesn't affect the product build or source build. These builds are the "old way" of shipping the .NET Core product: the dependencies "roll up" in a series of isolated builds to form the full product.
+
+*Product build*: a build that produced a *full product* for all platforms Microsoft ships.
+
+*Source build*: a build that doesn't depend on any external pre-built dependencies to produce a *full product* for a specific platform. AKA "build from source".
+
+*Orchestrator*: the system that keeps track of all repos involved in a build and performs the steps necessary to coordinate the build process. This doc assumes the orchestrator has some control over the machine(s) being built on: at minimum, downloading files to disk and customizing build command args. The orchestrator is involved in source builds and product builds, but not the isolated builds.
+
+*Repo build*: a particular repository's build that the orchestrator orchestrates. If building the repo involves multiple build legs (e.g. to build native bits on different platforms), all legs together are called *a* repo build.
+
+*Blob feed*: a structured directory of build outputs. AKA "transport feed". Summary of important details for auto dependency flow:
+
+ * There is one blob feed per product build. Repo builds add to the single blob feed.
+ * The orchestrator makes the blob feed available to all repo builds.
+ * The blob feed is on disk or in blob storage virtual directories, but in general it has the same structure either way.
+
+*Upstream/Downstream*: when dicussing repo A, repo B is "upstream" if A has a dependency on product outputs from B. Only repos built by the orchestrator are included. For example, in this doc, external toolsets aren't considered upstream from any repository, nor are any redistributed libraries that we don't build ourselves.
+
+
+# Inserting dependency changes
+
+The dependency change flow is shared by all builds that involve the orchestrator.
+
+TL;DR: the orchestrator derives package version props from the blob feed after each repo build and feeds the data into all subsequent repo builds.
+
+## Vision: product build end-to-end
+This narrative describes an end-to-end Core-Setup (Runtime) product build using CoreFX, CoreCLR, and Standard. More repos will be involved in real product builds.
+
+The orchestrator starts by calculating a repo dependency graph. Edges are determined by comparing each repo's *repo output manifest* and *repo input manifest*. The graph is a DAG, so it's buildable. In this case, we have:
```

Comment:
> repo's *repo output manifest* and *repo input manifest*

I question how dynamic this graph needs to be.  In my mind, I don't see the dependencies between repos changing often enough to warrant a dynamic system where the orchestrator is completely agnostic of the dependencies.  I think the system we have in source-build today works now, and will work for the foreseeable future.

I think the amount of work to build and maintain these input/output manifests for each repo will not be worth the value gained.

Reply:
I see this is mentioned in the "Incremental steps toward the vision" section below.  Do we see this as necessary even in the "long term" vision?

Reply:
I suppose that depends on how hard it is to change the dependencies later. If the graph is easily updated by checking into the orchestrator itself, it might not be worth the effort. Also, it might be easier to see problems if there's one source of truth for the graph rather than having it sprinkled around the repos.

Reply:
I agree that keeping the "orchestrator" as simple as possible is important.  Clear declarative contracts are good, and having a ridiculously straight forward orchestrator initially will allow the build to stand up, and then we'll really know what needed and where the highest priority problems are

Reply:
I'm going to add a commit removing the "far enough forward we probably won't ever do it" stuff so we have a concrete idea what that looks like. (I also prefer removing it at this point.)

@markwilkie, what do you mean by "Clear declarative contracts are good"? That seems to argue the opposite of this thread, which is proposing to *remove* a potential declarative contract and instead (continue to) store the metadata in the source-build repository.

Reply:
It seems to me that declaring "produces/consumes" as a list in each repo is a good thing.  However, this doesn't extend to a graph building fancy pants orchestrator.  (grin)

Reply:
It still sounds to me like you're arguing for these contracts, but in this context, `repo output/input manifest == produces/consumes`, and I'm getting rid of them. The only use case for repo output/input is removing the "which repo builds first" metadata from source-build, and instead associating repos based on what packages they output/input.

The orchestrator has to build a graph either way, and IMO it's only marginally more challenging with repo input/output manifests. The problem is *asking repos to produce this information without first performing a build*: this is a rather large implementation/maintenance cost for an undetermined--almost definitely small--gain. (Even after performing a build, this information is not necessarily easy to gather.)

Reply:
One of the things we learned with desktop is that it's becomes very expensive to manage a product/build when it's unclear what the dependencies are.  (it's becomes tough to change anything)  The core "requirement" here (for me anyway) is that there's a reasonable way to manage the build over time.  A proven strategy is to leverage contracts to help do this.  (no magic bullets of course....)

I suspect there are *many* implementations that will work.  For example, where the contract (input/output list?) lives is probably not important.  What's important (I think) is that it's explicit what the dependencies are and not simply implied as a "side effect" of the build scripts themselves.  Knowing the implications what one's change in that world is a tough place to live.

Reply:
I see--you could say that each repo `produces` itself, and each repo in source-build has a list of repos that it `consumes`. The contract for `produces` is identity, and for `consumes` it's having a `RepositoryReference` item of another repo's name.

Reply:
Perhaps, although referencing packages is probably the most concrete (and known) way to take these dependencies.  So long as we're not taking crazy dependencies on specific little packages, I presume this is a reasonable thing to do.  (but I don't know that for sure)

Reply:
> I presume this is a reasonable thing to do.

If "this" is "have repos list the packages they depend on and produce before any building happens", the consensus so far is that it's not reasonable and we shouldn't ask repos to do it. Instead, we expect that repo-level dependencies defined in the source-build repo are clear and easy enough to maintain that we should stick with that. (What I described above is the current state of the repo's "contracts", which can also be used by the product build orchestrator.)